internal/worker/concurrency_auto_scaler.go

```diff
@@ -0,0 +1,305 @@
+	if !c.enabled { // skip update if auto scaler is disabled
+		c.logEvent(autoScalerEventPollerSkipUpdateNotEnabled)
+		return
+	}
+	updateTime := c.clock.Now()
+	if updateTime.Before(c.pollerPermitLastUpdate.Add(c.cooldown)) { // before cooldown
+		c.logEvent(autoScalerEventPollerSkipUpdateCooldown)
+		return
+	}
+
+	newQuota := c.concurrency.PollerPermit.Quota()
+	pollerWaitTime := c.pollerWaitTime.Average()
+	if pollerWaitTime < lowerPollerWaitTime { // pollers are busy
+		newQuota = c.scaleUpPollerPermit(pollerWaitTime)
+	} else if pollerWaitTime > upperPollerWaitTime { // pollers are idle
+		newQuota = c.scaleDownPollerPermit(pollerWaitTime)
+	} else {
+		c.logEvent(autoScalerEventPollerSkipUpdateNoChange)
+		return
+	}
+	c.concurrency.PollerPermit.SetQuota(newQuota)
+	c.pollerPermitLastUpdate = updateTime
+	c.logEvent(autoScalerEventPollerUpdate)
+}
+
+func (c *ConcurrencyAutoScaler) scaleUpPollerPermit(pollerWaitTime time.Duration) int {
+	currentQuota := c.concurrency.PollerPermit.Quota()
+
+	// inverse scaling with edge case of 0 wait time
+	// use logrithm to smooth the scaling to avoid drastic change
+	newQuota := math.Round(
+		float64(currentQuota) * smoothingFunc(lowerPollerWaitTime) / smoothingFunc(pollerWaitTime))
+	newQuota = math.Max(
+		float64(c.pollerMinCount),
+		math.Min(float64(c.pollerMaxCount), newQuota),
+	)
+	return int(newQuota)
+}
+
+func (c *ConcurrencyAutoScaler) scaleDownPollerPermit(pollerWaitTime time.Duration) int {
+	currentQuota := c.concurrency.PollerPermit.Quota()
+
+	// inverse scaling with edge case of 0 wait time
+	// use logrithm to smooth the scaling to avoid drastic change
+	newQuota := math.Round(
+		float64(currentQuota) * smoothingFunc(upperPollerWaitTime) / smoothingFunc(pollerWaitTime))
+	newQuota = math.Max(
+		float64(c.pollerMinCount),
+		math.Min(float64(c.pollerMaxCount), newQuota),
+	)
+	return int(newQuota)
+}
+
+// smoothingFunc is a log2 function with offset to smooth the scaling and address 0 values
+func smoothingFunc(x time.Duration) float64 {
+	return math.Log2(2 + float64(x/time.Millisecond))
+}
+
+type number interface {
+	int64 | float64 | time.Duration
+}
+
+type rollingAverage[T number] struct {
+	mu     sync.RWMutex
+	window []T
+	index  int
+	sum    T
+	count  int
+}
+
+func newRollingAverage[T number](capacity int) *rollingAverage[T] {
+	return &rollingAverage[T]{
+		window: make([]T, capacity),
+	}
+}
+
+// Add always add positive numbers
+func (r *rollingAverage[T]) Add(value T) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// no op on zero rolling window
+	if len(r.window) == 0 {
+		return
+	}
+
+	// replace the old value with the new value
+	r.sum += value - r.window[r.index]
+	r.window[r.index] = value
+	r.index++
+	r.index %= len(r.window)
+
+	if r.count < len(r.window) {
+		r.count++
+	}
+}
+
+func (r *rollingAverage[T]) Average() T {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	if r.count == 0 {
+		return 0
+	}
+	return r.sum / T(r.count)
```

Comment:
are there any conditions under which sum might become zero?

Reply:
Yes, if Average() is called before any `Add` method is called, the returned value will be zero. For usages of this rollingAverage, it needs to handle zero value case. 
In auto scaler, I handled it through a `SmoothFunc`